embedded/store/tx.go

```diff
@@ -209,48 +222,67 @@ func (hdr *TxHeader) innerHash() [sha256.Size]byte {
 	copy(b[i:], hdr.BlRoot[:])
 	i += sha256.Size
 
-	// hash(ts + mdLen + md + nentries + eH + blTxID + blRoot)
+	// hash(ts + version + (mdLen + md) + nentries + eH + blTxID + blRoot)
 	return sha256.Sum256(b[:i])
 }
 
+// Alh calculates the Accumulative Linear Hash up to this transaction
+// Alh is calculated as hash(txID + prevAlh + hash(ts + nentries + eH + blTxID + blRoot))
+// Inner hash is calculated so to reduce the length of linear proofs
 func (hdr *TxHeader) Alh() [sha256.Size]byte {
 	// txID + prevAlh + innerHash
 	var bi [txIDSize + 2*sha256.Size]byte
 	binary.BigEndian.PutUint64(bi[:], hdr.ID)
 	copy(bi[txIDSize:], hdr.PrevAlh[:])
 
-	// hash(ts + mdLen + md + nentries + eH + blTxID + blRoot)
+	// hash(ts + version + (mdLen + md)? + nentries + eH + blTxID + blRoot)
 	innerHash := hdr.innerHash()
 	copy(bi[txIDSize+sha256.Size:], innerHash[:])
 
 	// hash(txID + prevAlh + innerHash)
 	return sha256.Sum256(bi[:])
 }
 
+func (tx *Tx) TxEntryDigest() (TxEntryDigest, error) {
+	switch tx.header.Version {
+	case 0:
+		return TxEntryDigest_v1_1, nil
+	case 1:
+		return TxEntryDigest_v1_2, nil
+	}
+
+	return nil, ErrCorruptedData
+}
+
 func (tx *Tx) BuildHashTree() error {
-	digests := make([][sha256.Size]byte, tx.nentries)
+	digests := make([][sha256.Size]byte, tx.header.NEntries)
+
+	txEntryDigest, err := tx.TxEntryDigest()
+	if err != nil {
+		return err
+	}
 
 	for i, e := range tx.Entries() {
-		digests[i] = e.Digest()
+		digests[i] = txEntryDigest(e)
 	}
 
-	return tx.htree.BuildWith(digests)
-}
+	err = tx.htree.BuildWith(digests)
+	if err != nil {
+		return err
+	}
 
-func (tx *Tx) Entries() []*TxEntry {
-	return tx.entries[:tx.nentries]
-}
+	root, err := tx.htree.Root()
+	if err != nil {
+		return err
+	}
 
-// Alh calculates the Accumulative Linear Hash up to this transaction
-// Alh is calculated as hash(txID + prevAlh + hash(ts + nentries + eH + blTxID + blRoot))
-// Inner hash is calculated so to reduce the length of linear proofs
-func (tx *Tx) CalcAlh() {
-	tx.Alh = tx.Header().Alh()
+	tx.header.Eh = root
```

Comment:
That's something new compared to the previous body of the function - this is about fixing the bug of incorrect htree root in transaction header we discussed before?

Reply:
it's the same behaviour but refactored when introducing the header struct into the tx. The hash of the tx is directly calculated from the header, the root hash of the entries is set when the hash tree is calculated.